timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/TimelockLeadershipMetrics.java

```diff
@@ -16,58 +16,120 @@
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
 import com.palantir.tritium.metrics.registry.MetricName;
 
 @Value.Immutable
 public abstract class TimelockLeadershipMetrics {
 
-    @Value.Parameter
-    abstract TimelockPaxosMetrics metrics();
+    public abstract TimelockPaxosMetrics metrics();
+    public abstract UUID leaderUuid();
+    public abstract PingableLeader localPingableLeader();
+    public abstract AutobatchingLeadershipObserverFactory leadershipObserverFactory();
+    abstract Client client();
 
-    @Value.Parameter
-    abstract PingableLeader localPingableLeader();
+    @Value.Derived
+    List<SafeArg<String>> namespaceAsArgs() {
+        return ImmutableList.of(
+                SafeArg.of(TAG_PAXOS_USE_CASE, metrics().paxosUseCase().toString()),
+                SafeArg.of(TAG_CLIENT, client().value()));
+    }
 
     @Value.Derived
-    MetricsManager metricsManager() {
-        // we don't use the normal metric registry so we don't care about this
-        return MetricsManagers.of(new MetricRegistry(), metrics().metrics());
+    public PaxosLeadershipEventRecorder eventRecorder() {
+        return PaxosLeadershipEventRecorder.create(
+                metrics().metrics(),
+                leaderUuid().toString(),
+                leadershipObserver(),
+                namespaceAsArgs());
     }
 
-    public <T> T instrument(Client client, String name, Class<T> clazz, T instance) {
+    @Value.Derived
+    LeadershipObserver leadershipObserver() {
+        return leadershipObserverFactory().create(client());
+    }
+
+    public <T> T instrument(String name, Class<T> clazz, T instance) {
         return AtlasDbMetrics.instrumentWithTaggedMetrics(
                 metrics().metrics(),
                 clazz,
                 instance,
                 name,
                 _context -> ImmutableMap.of(
-                        AtlasDbMetricNames.TAG_CLIENT, client.value(),
-                        AtlasDbMetricNames.TAG_CURRENT_SUSPECTED_LEADER, String.valueOf(localPingableLeader().ping())));
+                        TAG_CLIENT, client().value(),
+                        TAG_CURRENT_SUSPECTED_LEADER, String.valueOf(localPingableLeader().ping())));
     }
 
-    @Value.Derived
-    public List<SafeArg<Object>> namespaceAsArgs() {
-        return ImmutableList.of(SafeArg.of(AtlasDbMetricNames.TAG_PAXOS_USE_CASE, metrics().paxosUseCase().toString()));
+    public static AutobatchingLeadershipObserverFactory createFactory(TimelockPaxosMetrics metrics) {
+        return AutobatchingLeadershipObserverFactory.create(getMetricDeregistrator(metrics));
+    }
+
+    private static Consumer<SetMultimap<LeadershipEvent, Client>> getMetricDeregistrator(TimelockPaxosMetrics metrics) {
+        PaxosUseCase paxosUseCase = metrics.paxosUseCase();
+        switch (paxosUseCase) {
+            case LEADER_FOR_ALL_CLIENTS:
+                return deregisterMetricsForSingleLeader(metrics);
```

Comment:
this would be a perfect use for derive4j, but alas that is a refactoring for another day